sdk/go/common/util/mapper/mapper_test.go

```diff
@@ -106,34 +106,131 @@ func TestFieldMapper(t *testing.T) {
 }
 
 type bagtag struct {
-	String        string `pulumi:"s"`
-	StringSkip    string `pulumi:"sc,skip"`
-	StringOpt     string `pulumi:"so,optional"`
-	StringSkipOpt string `pulumi:"sco,skip,optional"`
-}
-
-type AnInterface interface {
-	isAnInterface()
+	String        string                 `pulumi:"s"`
+	StringSkip    string                 `pulumi:"sc,skip"`
+	StringOpt     string                 `pulumi:"so,optional"`
+	StringSkipOpt string                 `pulumi:"sco,skip,optional"`
+	MapOpt        map[string]interface{} `pulumi:"mo,optional"`
 }
 
 func TestMapperEncode(t *testing.T) {
 	t.Parallel()
 	bag := bagtag{
 		String:    "something",
 		StringOpt: "ohmv",
+		MapOpt: map[string]interface{}{
+			"a": "something",
+			"b": nil,
+		},
 	}
 
-	md := New(nil)
-	m, err := md.Encode(bag)
+	md := &mapper{}
+	var err error
+	var m map[string]interface{}
+
+	// Nils
+	m, err = md.Encode(nil)
+	require.NoError(t, err)
+	assert.Len(t, m, 0)
+
+	// Structs
+	m, err = md.encode(reflect.ValueOf(bag))
+	require.NoError(t, err)
+	assert.Equal(t, "something", m["s"])
+	assert.Equal(t, "ohmv", m["so"])
+	assert.Equal(t, map[string]interface{}{"a": "something", "b": nil}, m["mo"])
+
+	// Pointers
+	m, err = md.encode(reflect.Zero(reflect.TypeOf(&bag)))
+	require.NoError(t, err)
+	assert.Nil(t, m)
+	m, err = md.encode(reflect.ValueOf(&bag))
 	require.NoError(t, err)
 	assert.Equal(t, "something", m["s"])
 	assert.Equal(t, "ohmv", m["so"])
+	assert.Equal(t, map[string]interface{}{"a": "something", "b": nil}, m["mo"])
+}
+
+func TestMapperEncodeValue(t *testing.T) {
+	t.Parallel()
+	strdata := "something"
+	bag := bagtag{
+		String:    "something",
+		StringOpt: "ohmv",
+	}
+	slice := []string{"something"}
+	mapdata := map[string]interface{}{
+		"a": "something",
+		"b": nil,
+	}
+	anyType := reflect.TypeOf((*any)(nil)).Elem()
+	assert.Equal(t, reflect.Interface, anyType.Kind())
 
-	// Encode a nil interface
+	md := &mapper{}
+	var err error
+	var v any
 
-	m, err = md.Encode((AnInterface)(nil))
```

Comment:
nil and a nil interface show up different in reflect so I think we still want this test case.

Reply:
It is not a meaningful distinction because the public `Encode` function performs a null check before using reflection:
```go
func (md *mapper) Encode(source interface{}) (map[string]interface{}, MappingError) {
	if source == nil {
		return nil, nil
	}
	return md.encode(reflect.ValueOf(source))
}
```

Of course, the distinction is more significant after you have a `reflect.Value` and all the cases are covered in `TestMapperEncodeValue`.  What do you think?

Reply:
Probably still safer to have that test case just in case anyone comes and moves things around in Encode, would be good to ensure we don't screw up when fat pointers get encountered.